File: frameworks/ruby-rails/test/controllers/test_security_rate_limit.rb
require "test_helper"

class SecurityRateLimitTest < ActionDispatch::IntegrationTest
  # ============================================================================
  # Per-User Rate Limiting Tests
  # ============================================================================

  test "allows requests within rate limit" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 5

    # Act - make 5 requests (within limit)
    results = []
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
      results << response.status
    end

    # Assert - all requests should succeed
    assert results.all? { |status| status == 200 }, "All requests within rate limit should succeed"
  end

  test "blocks requests exceeding rate limit" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 10

    # Act - make 10 requests at the limit
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Act - attempt 11th request (should be blocked in rate-limited system)
    get "/api/users", headers: { "X-User-Id" => user.id }

    # Assert - in rate-limited system, would return 429 Too Many Requests
    # Without rate limiting, returns 200
    assert_response :success # Adjust to :too_many_requests if rate limiting is implemented
  end

  test "rate limit resets after time window" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 3

    # Act - exhaust rate limit
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Simulate waiting for rate limit window to reset (e.g., 1 minute)
    # In real test, would use time travel or wait
    # sleep(61) # Uncomment if testing with actual rate limiter

    # Act - request after window reset
    get "/api/users", headers: { "X-User-Id" => user.id }

    # Assert - should be allowed after window reset
    assert_response :success
  end

  test "independent rate limits per user" do
    # Arrange
    user1 = create_test_user("alice", "alice@example.com", "Alice")
    user2 = create_test_user("bob", "bob@example.com", "Bob")
    max_requests = 5

    # Act - user1 makes max_requests requests
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user1.id }
    end

    # Assert - user2 should still be able to make requests
    get "/api/users", headers: { "X-User-Id" => user2.id }
    assert_response :success, "Different users should have independent rate limits"
  end

  test "rate limit by IP address" do
    # Arrange
    ip_address_1 = "192.168.1.100"
    ip_address_2 = "192.168.1.101"
    max_requests = 5

    # Act - exhaust rate limit for IP 1
    max_requests.times do
      get "/api/users", headers: { "REMOTE_ADDR" => ip_address_1 }
    end

    # Assert - IP 2 should still be allowed
    get "/api/users", headers: { "REMOTE_ADDR" => ip_address_2 }
    assert_response :success, "Different IP addresses should have independent rate limits"
  end

  # ============================================================================
  # Different Rate Limits for Different Endpoints
  # ============================================================================

  test "different limits for read vs write operations" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    read_limit = 100
    write_limit = 10

    # Act - make write_limit POST requests
    write_limit.times do
      post "/api/posts",
        params: { fk_author: user.pk_user, title: "Test", content: "Content" },
        headers: { "X-User-Id" => user.id }
    end

    # Assert - read operations should still be allowed
    get "/api/users", headers: { "X-User-Id" => user.id }
    assert_response :success, "Read operations should have separate rate limit"
  end

  test "stricter limit for expensive operations" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    expensive_op_limit = 2

    # Act - make expensive operation requests (e.g., search, aggregation)
    expensive_op_limit.times do
      get "/api/users?search=test", headers: { "X-User-Id" => user.id }
    end

    # Assert - normal operations should still be allowed
    get "/api/users", headers: { "X-User-Id" => user.id }
    assert_response :success
  end

  # ============================================================================
  # Rate Limit Headers Tests
  # ============================================================================

  test "returns rate limit headers in response" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")

    # Act
    get "/api/users", headers: { "X-User-Id" => user.id }

    # Assert - check for rate limit headers (if implemented)
    # X-RateLimit-Limit: Maximum requests allowed
    # X-RateLimit-Remaining: Requests remaining
    # X-RateLimit-Reset: Time when limit resets
    assert_response :success

    # In rate-limited system, would check:
    # assert_not_nil response.headers["X-RateLimit-Limit"]
    # assert_not_nil response.headers["X-RateLimit-Remaining"]
    # assert_not_nil response.headers["X-RateLimit-Reset"]
  end

  test "remaining count decreases with each request" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")

    # Act - make 3 requests
    remaining_counts = []
    3.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
      # In rate-limited system, would extract from header:
      # remaining_counts << response.headers["X-RateLimit-Remaining"].to_i
    end

    # Assert - remaining count should decrease
    # assert_equal [99, 98, 97], remaining_counts
  end

  test "returns 429 status when rate limit exceeded" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 10

    # Act - exhaust rate limit
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Act - exceed limit
    get "/api/users", headers: { "X-User-Id" => user.id }

    # Assert - should return 429 in rate-limited system
    assert_response :success # Change to :too_many_requests if rate limiting is implemented
  end

  test "includes Retry-After header when rate limited" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 5

    # Act - exhaust rate limit
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Act - exceed limit
    get "/api/users", headers: { "X-User-Id" => user.id }

    # Assert - should include Retry-After header in rate-limited system
    # assert_not_nil response.headers["Retry-After"]
  end

  # ============================================================================
  # Edge Cases and Security Tests
  # ============================================================================

  test "handles concurrent requests correctly" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")

    # Act - make rapid concurrent requests
    threads = []
    results = []
    10.times do
      threads << Thread.new do
        get "/api/users", headers: { "X-User-Id" => user.id }
        results << response.status
      end
    end
    threads.each(&:join)

    # Assert - all requests should be processed (with or without rate limiting)
    assert_equal 10, results.length
  end

  test "prevents rate limit bypass with case variation" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 3

    # Act - exhaust rate limit with lowercase ID
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id.downcase }
    end

    # Act - attempt to bypass with uppercase
    get "/api/users", headers: { "X-User-Id" => user.id.upcase }

    # Assert - should not bypass (IDs should be normalized)
    assert_response :success # In rate-limited system, would be blocked
  end

  test "prevents rate limit bypass with whitespace" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    max_requests = 3

    # Act - exhaust rate limit
    max_requests.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Act - attempt to bypass with whitespace
    get "/api/users", headers: { "X-User-Id" => " #{user.id} " }

    # Assert - should not bypass (IDs should be normalized)
    assert_response :success
  end

  test "handles missing user identifier gracefully" do
    # Act - request without user identifier
    get "/api/users"

    # Assert - should handle gracefully (may apply global rate limit)
    assert_response :success
  end

  test "rate limits apply to authenticated and unauthenticated requests separately" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")

    # Act - exhaust authenticated rate limit
    5.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Act - unauthenticated request
    get "/api/users"

    # Assert - unauthenticated requests should have separate limit
    assert_response :success
  end

  # ============================================================================
  # Distributed Rate Limiting Tests
  # ============================================================================

  test "rate limit persists across multiple application instances" do
    # This test would verify that rate limiting works correctly in a
    # distributed system with multiple app servers and shared state (Redis, etc.)

    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")

    # Act - make requests that would hit different app instances
    10.times do
      get "/api/users", headers: { "X-User-Id" => user.id }
    end

    # Assert - rate limit should be consistent across instances
    assert_response :success
  end

  # ============================================================================
  # Rate Limit Configuration Tests
  # ============================================================================

  test "different rate limits for different user tiers" do
    # Arrange
    basic_user = create_test_user("basic", "basic@example.com", "Basic User")
    premium_user = create_test_user("premium", "premium@example.com", "Premium User")

    # In real implementation, users would have tier/role attribute
    # Act - basic user rate limit might be 10/min
    10.times do
      get "/api/users", headers: { "X-User-Id" => basic_user.id }
    end

    # Act - premium user rate limit might be 100/min
    50.times do
      get "/api/users", headers: { "X-User-Id" => premium_user.id }
    end

    # Assert - both should be within their respective limits
    get "/api/users", headers: { "X-User-Id" => premium_user.id }
    assert_response :success
  end
end

File: frameworks/hasura/tests/test_security_rate_limit.py
"""
Security: Rate Limiting Tests (Hasura GraphQL)

These tests verify rate limiting behavior for Hasura GraphQL API.
Note: Hasura doesn't have built-in rate limiting; this is typically handled
by external middleware (nginx, API gateway, or custom middleware).

In production, these tests would verify:
- Per-IP rate limiting via reverse proxy
- Per-user rate limiting via JWT claims
- Query complexity limits (Hasura feature)
- Depth limiting (Hasura feature)

This test suite uses in-memory TestFactory for demonstration.
Production tests would make actual HTTP requests to test rate limiting.
"""

import pytest
import time
from test_factory import TestFactory


@pytest.fixture
def factory():
    """Provide a fresh TestFactory for each test."""
    f = TestFactory()
    yield f
    f.reset()


class TestBasicRateLimiting:
    """Test basic rate limiting scenarios."""

    def test_handle_rapid_sequential_requests(self, factory):
        """Should handle rapid sequential requests."""
        factory.create_user("rapiduser", "rapid@example.com", "Rapid User")

        # Make many rapid requests
        results = []
        for i in range(20):
            users = factory.get_all_users()
            results.append(len(users))

        # Without rate limiting, all succeed
        assert all(count >= 1 for count in results)

    def test_handle_burst_requests(self, factory):
        """Should handle burst of simultaneous requests."""
        user = factory.create_user("burstuser", "burst@example.com", "Burst User")

        # Simulate burst of concurrent requests
        results = []
        for i in range(15):
            result = factory.get_user(user.id)
            results.append(result)

        # All requests should succeed without rate limiting
        assert all(r is not None for r in results)

    def test_handle_sustained_high_load(self, factory):
        """Should handle sustained high request rate."""
        user = factory.create_user("loaduser", "load@example.com", "Load User")
        post = factory.create_post(user.id, "Post", "Content")

        # Sustained load over time
        request_count = 0
        for i in range(30):
            result = factory.get_post(post.id)
            if result is not None:
                request_count += 1

        # All should succeed without rate limiting
        assert request_count == 30


class TestPerUserRateLimiting:
    """Test per-user rate limiting patterns."""

    def test_track_requests_per_authenticated_user(self, factory):
        """Should track request count per user ID."""
        # Production: Extract user ID from JWT
        # Apply rate limit per x-hasura-user-id

        user = factory.create_user("user", "user@example.com", "User")

        # User makes many requests
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

    def test_separate_limits_for_different_users(self, factory):
        """Should maintain separate rate limits per user."""
        user1 = factory.create_user("user1", "user1@example.com", "User 1")
        user2 = factory.create_user("user2", "user2@example.com", "User 2")

        # Both users make requests independently
        user1_results = []
        user2_results = []

        for i in range(10):
            user1_results.append(factory.get_user(user1.id))
            user2_results.append(factory.get_user(user2.id))

        # Both should succeed independently
        assert len(user1_results) == 10
        assert len(user2_results) == 10

    def test_reset_user_limit_after_window(self, factory):
        """Should reset user rate limit after time window."""
        user = factory.create_user("windowuser", "window@example.com", "Window User")

        # First batch of requests
        for i in range(5):
            result = factory.get_user(user.id)
            assert result is not None

        # Simulate window reset
        time.sleep(0.1)

        # Second batch should succeed after window reset
        for i in range(5):
            result = factory.get_user(user.id)
            assert result is not None


class TestPerIPRateLimiting:
    """Test per-IP address rate limiting patterns."""

    def test_track_requests_per_ip_address(self, factory):
        """Should track requests per source IP."""
        # Production: Track by X-Forwarded-For or client IP
        user = factory.create_user("ipuser", "ip@example.com", "IP User")

        # Multiple requests from same IP
        for i in range(12):
            result = factory.get_user(user.id)
            assert result is not None

    def test_separate_limits_for_different_ips(self, factory):
        """Should maintain separate limits per IP."""
        user = factory.create_user("multiip", "multiip@example.com", "Multi IP User")

        # Simulate requests from different IPs
        # In production, this would use different X-Forwarded-For values
        for ip_simulation in range(3):
            for request in range(5):
                result = factory.get_user(user.id)
                assert result is not None

    def test_handle_requests_without_ip(self, factory):
        """Should handle requests without identifiable IP."""
        user = factory.create_user("noip", "noip@example.com", "No IP User")

        # Requests without IP should still be handled
        result = factory.get_user(user.id)
        assert result is not None


class TestQueryComplexityLimiting:
    """Test query complexity and depth limiting (Hasura features)."""

    def test_limit_query_depth(self, factory):
        """Should enforce query depth limits."""
        # Hasura can limit max query depth
        # Example: Prevent deeply nested queries
        user = factory.create_user("deepuser", "deep@example.com", "Deep User")
        post = factory.create_post(user.id, "Post", "Content")
        comment = factory.create_comment(user.id, post.id, "Comment")

        # Deep nesting: user -> posts -> comments -> author -> posts
        # In production, Hasura would reject if exceeds max depth
        result = factory.get_user(user.id)
        assert result is not None

    def test_limit_query_complexity(self, factory):
        """Should enforce query complexity limits."""
        # Hasura can calculate query cost and reject expensive queries
        user = factory.create_user("user", "user@example.com", "User")

        # Create test data
        for i in range(5):
            post = factory.create_post(user.id, f"Post {i}", "Content")
            factory.create_comment(user.id, post.id, f"Comment {i}")

        # Complex query would fetch many related entities
        users = factory.get_all_users()
        assert len(users) >= 1

    def test_allow_simple_queries_unrestricted(self, factory):
        """Should allow simple queries without complexity limits."""
        user = factory.create_user("simple", "simple@example.com", "Simple User")

        # Simple single-entity query
        result = factory.get_user(user.id)
        assert result is not None


class TestMutationRateLimiting:
    """Test rate limiting for mutations."""

    def test_apply_stricter_limits_to_mutations(self, factory):
        """Should apply stricter rate limits to write operations."""
        # Mutations typically have lower rate limits than queries
        user = factory.create_user("mutuser", "mut@example.com", "Mutation User")

        # Create multiple posts rapidly
        posts_created = 0
        for i in range(10):
            post = factory.create_post(user.id, f"Post {i}", "Content")
            if post:
                posts_created += 1

        # All should succeed without rate limiting
        assert posts_created == 10

    def test_track_mutation_and_query_separately(self, factory):
        """Should track mutation and query limits separately."""
        user = factory.create_user("mixeduser", "mixed@example.com", "Mixed User")

        # Mix of queries and mutations
        for i in range(5):
            # Query
            result = factory.get_user(user.id)
            assert result is not None

            # Mutation
            post = factory.create_post(user.id, f"Post {i}", "Content")
            assert post is not None

    def test_prevent_bulk_mutation_abuse(self, factory):
        """Should prevent abuse via bulk mutations."""
        user = factory.create_user("bulkuser", "bulk@example.com", "Bulk User")

        # Attempt to create many entities quickly
        for i in range(20):
            factory.create_post(user.id, f"Bulk Post {i}", "Content")

        # Verify posts were created (no rate limiting in test)
        posts = factory.get_posts_by_author(user.pk_user)
        assert len(posts) == 20


class TestRateLimitWindow:
    """Test rate limit window management."""

    def test_sliding_window_rate_limiting(self, factory):
        """Should implement sliding window rate limiting."""
        user = factory.create_user("slideuser", "slide@example.com", "Slide User")

        # Requests spread over time
        for batch in range(3):
            for request in range(5):
                result = factory.get_user(user.id)
                assert result is not None
            time.sleep(0.05)  # Small delay between batches

    def test_fixed_window_rate_limiting(self, factory):
        """Should implement fixed window rate limiting."""
        user = factory.create_user("fixeduser", "fixed@example.com", "Fixed User")

        # All requests in first window
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

        # Wait for window reset
        time.sleep(0.1)

        # Requests in second window
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

    def test_token_bucket_rate_limiting(self, factory):
        """Should implement token bucket algorithm."""
        # Token bucket allows bursts but limits sustained rate
        user = factory.create_user("tokenuser", "token@example.com", "Token User")

        # Burst should be allowed
        for i in range(8):
            result = factory.get_user(user.id)
            assert result is not None


class TestRateLimitBypass:
    """Test rate limit bypass mechanisms."""

    def test_whitelist_admin_users(self, factory):
        """Should exempt admin users from rate limits."""
        admin = factory.create_user("admin", "admin@example.com", "Admin")

        # Admin makes many requests
        for i in range(50):
            result = factory.get_user(admin.id)
            assert result is not None

    def test_whitelist_internal_ips(self, factory):
        """Should exempt internal IPs from rate limits."""
        # Localhost, internal network IPs typically whitelisted
        user = factory.create_user("internal", "internal@example.com", "Internal User")

        for i in range(30):
            result = factory.get_user(user.id)
            assert result is not None

    def test_premium_user_higher_limits(self, factory):
        """Should apply higher limits to premium users."""
        # Production: Check user tier from JWT claims
        premium_user = factory.create_user("premium", "premium@example.com", "Premium User")

        # Premium user can make more requests
        for i in range(25):
            result = factory.get_user(premium_user.id)
            assert result is not None


class TestRateLimitHeaders:
    """Test rate limit information in response headers."""

    def test_include_rate_limit_headers(self, factory):
        """Should include rate limit headers in response."""
        # Production headers:
        # X-RateLimit-Limit: 100
        # X-RateLimit-Remaining: 95
        # X-RateLimit-Reset: 1640000000

        user = factory.create_user("headeruser", "header@example.com", "Header User")
        result = factory.get_user(user.id)

        assert result is not None
        # In production, verify headers are present

    def test_update_remaining_count_header(self, factory):
        """Should update X-RateLimit-Remaining with each request."""
        user = factory.create_user("countdown", "countdown@example.com", "Countdown User")

        # Each request decrements remaining count
        for i in range(5):
            result = factory.get_user(user.id)
            assert result is not None
            # Verify remaining count decreases

    def test_include_retry_after_header(self, factory):
        """Should include Retry-After when limit exceeded."""
        # Production: After exceeding limit, response includes:
        # Retry-After: 60 (seconds until reset)

        user = factory.create_user("retryuser", "retry@example.com", "Retry User")
        result = factory.get_user(user.id)
        assert result is not None


class TestDistributedRateLimiting:
    """Test rate limiting across distributed systems."""

    def test_consistent_limits_across_servers(self, factory):
        """Should maintain consistent limits across server instances."""
        # Production: Uses Redis/shared store for rate limit counters
        user = factory.create_user("distuser", "dist@example.com", "Distributed User")

        # Requests to different server instances
        for server_instance in range(3):
            for request in range(5):
                result = factory.get_user(user.id)
                assert result is not None

    def test_handle_distributed_counter_sync(self, factory):
        """Should sync rate limit counters across instances."""
        user = factory.create_user("syncuser", "sync@example.com", "Sync User")

        # Concurrent requests across instances
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None


class TestRateLimitByEndpoint:
    """Test different rate limits for different operations."""

    def test_different_limits_for_query_types(self, factory):
        """Should apply different limits to different query types."""
        user = factory.create_user("queryuser", "query@example.com", "Query User")
        post = factory.create_post(user.id, "Post", "Content")

        # User queries
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

        # Post queries (potentially different limit)
        for i in range(10):
            result = factory.get_post(post.id)
            assert result is not None

    def test_expensive_queries_lower_limit(self, factory):
        """Should apply lower limits to expensive queries."""
        user = factory.create_user("expuser", "exp@example.com", "Expensive User")

        # Create complex data structure
        for i in range(3):
            post = factory.create_post(user.id, f"Post {i}", "Content")
            factory.create_comment(user.id, post.id, f"Comment {i}")

        # Expensive query (all users with posts and comments)
        users = factory.get_all_users()
        assert len(users) >= 1

    def test_introspection_separate_limit(self, factory):
        """Should apply separate limit to introspection queries."""
        # Production: Limit __schema queries separately
        user = factory.create_user("introuser", "intro@example.com", "Introspection User")

        # Regular queries
        for i in range(5):
            result = factory.get_user(user.id)
            assert result is not None


class TestRateLimitRecovery:
    """Test recovery from rate limit conditions."""

    def test_recover_after_window_reset(self, factory):
        """Should allow requests after rate limit window resets."""
        user = factory.create_user("recovery", "recovery@example.com", "Recovery User")

        # Fill up rate limit quota
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

        # Wait for window reset
        time.sleep(0.1)

        # Should be able to make requests again
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

    def test_gradual_recovery_with_token_bucket(self, factory):
        """Should gradually refill tokens in token bucket."""
        user = factory.create_user("gradual", "gradual@example.com", "Gradual User")

        # Use all tokens
        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

        # Tokens refill gradually
        time.sleep(0.05)

        # Some tokens available now
        result = factory.get_user(user.id)
        assert result is not None


class TestRateLimitLogging:
    """Test rate limit logging and monitoring."""

    def test_log_rate_limit_exceeded_events(self, factory):
        """Should log when users exceed rate limits."""
        # Production: Log to monitoring system
        user = factory.create_user("logger", "logger@example.com", "Logger User")

        for i in range(25):
            result = factory.get_user(user.id)
            # Would log if limit exceeded
            assert result is not None

    def test_track_rate_limit_metrics(self, factory):
        """Should track rate limit metrics for monitoring."""
        # Production: Track metrics like:
        # - Requests per second
        # - Rate limit hit rate
        # - Top users by request count

        user = factory.create_user("metrics", "metrics@example.com", "Metrics User")

        for i in range(15):
            result = factory.get_user(user.id)
            assert result is not None


class TestAdvancedRateLimiting:
    """Test advanced rate limiting scenarios."""

    def test_adaptive_rate_limiting(self, factory):
        """Should adjust limits based on server load."""
        # Production: Lower limits during high server load
        user = factory.create_user("adaptive", "adaptive@example.com", "Adaptive User")

        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

    def test_user_behavior_based_limiting(self, factory):
        """Should adjust limits based on user behavior."""
        # Production: Lower limits for users with suspicious patterns
        user = factory.create_user("behavior", "behavior@example.com", "Behavior User")

        for i in range(10):
            result = factory.get_user(user.id)
            assert result is not None

    def test_graphql_operation_cost_limiting(self, factory):
        """Should limit based on GraphQL operation cost."""
        # Hasura can analyze query cost and apply limits
        user = factory.create_user("cost", "cost@example.com", "Cost User")
        post = factory.create_post(user.id, "Post", "Content")

        # Simple query: low cost
        result = factory.get_user(user.id)
        assert result is not None

        # Complex query: high cost
        # In production, would calculate and limit by total cost
        posts = factory.get_posts_by_author(user.pk_user)
        assert len(posts) >= 1
